pyteal/ast/scratchvar.py

```diff
@@ -47,5 +54,110 @@ def load(self) -> ScratchLoad:
         """Load value from Scratch Space"""
         return self.slot.load(self.type)
 
+    def index(self) -> Expr:
+        return self.slot.index()
+
 
 ScratchVar.__module__ = "pyteal"
+
+
+class DynamicScratchVar:
```

Comment:
If you get rid of the `indexer` parameter, then I think `DynamicScratchVar` could inherit from `ScratchVar`.

This would be preferable because right now, the subroutine invocation code is "lying" to user-defined subroutine bodies, since they expect to be called with an instance of `ScratchVar` but are actually given a `DynamicScratchVar`. Making `DynamicScratchVar` a subclass of `ScratchVar` would fix this oddity.

Reply:
I'm going to push back a bit on this one, but I'm not dead-set against. I think the main issue is that I want to exclude using DynamicScratchVar's every where ScratchVar's are used. The best example is the one you just mentioned whereby I'm "lying". If a DynamicScratchVar is a ScratchVar, then it seems reasaonable to expect the users to write the subroutines with `: DynamicScratchVar`'s annotations instead. This is a can of worms I didn't want to pursue.

Reply:
We should be as truthful as possible with our type interfaces. If we say your subroutine will receive a `ScratchVar`, you should expect `isinstance(arg, ScratchVar)` to be true.

As for allowing `DynamicScratchVar` as a valid annotation on subroutines, I don't think we should allow this right now because I don't see any reason to do it. As far as pass-by-reference is concerned, `DynamicScratchVar` is just an implementation detail that's required to give subroutines a `ScratchVar`.

Reply:
As `DynamicScratchVar`'s are also intended for use directly (e.g. the Wilt Chamberlain snippet above) we can expect this to be used for more than just an implementation detail. I'll experiment with sub-classing and assess the impact.